### PR TITLE
3181 insert inbound return

### LINKED ISSUE #3181: API: insertInboundReturn
Epic #2897 

- Create an abstraction for StockIn lines
- Use this abstraction for inserting inbound returns
- Create a refactor issue to use this abstraction for Inbound

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #3181

# 👩🏻‍💻 What does this PR do? 
 

Adds API for inserting an inbound return.

Creates the StockIn abstraction, similar to StockOut. Uses this abstraction for inserting the inbound return. 

# 🧪 How has/should this change been tested? 

Automated tests added for service layer.

- From Outbound Shipment detail view (at least delivered status)
- Select lines and click the process return of selected lines action
- Set return quantities, and optionally reasons/notes
- Submit
- You'll get redirected, but the detail view isn't merged currently...
- You can query for the invoice in a graphql playground though!

### CODE REVIEW

File: client/packages/invoices/src/Returns/modals/InboundReturn/InboundReturn.tsx [hunk 1]
```diff
@@ -80,7 +78,7 @@ export const InboundReturnEditModal = ({
         }
         okButton={
           currentTab === Tabs.Reason ? (
-            <DialogButton onClick={onOk} variant="ok" disabled={!okEnabled} />
```
Comment (lache-melvin) [inline client/packages/invoices/src/Returns/modals/InboundReturn/InboundReturn.tsx:83]: we have reasons as optional in the backend, so have removed the constraint here 😅

File: client/packages/invoices/src/Returns/modals/ReturnReasonsTable.tsx [hunk 2]
```diff
@@ -46,6 +46,7 @@ export const ReturnReasonsComponent = ({
         Cell: TextInputCell,
         width: 200,
         setter: updateLine,
+        accessor: ({ rowData }) => rowData.note ?? '',
```
Comment (lache-melvin) [inline client/packages/invoices/src/Returns/modals/ReturnReasonsTable.tsx:49]: fixes a console warning (didn't like it being null)

File: client/packages/invoices/src/utils.ts [hunk 3]
```diff
@@ -226,6 +226,7 @@ export const canReturnInboundLines = (inbound: InboundFragment): boolean =>
   inbound.status === InvoiceNodeStatus.Verified;
 
 export const canReturnOutboundLines = (outbound: OutboundFragment): boolean =>
+  outbound.status === InvoiceNodeStatus.Shipped ||
```
Comment (lache-melvin) [inline client/packages/invoices/src/utils.ts:229]: If a manual outbound shipment, it never goes beyond shipped, but the supplier should be able to process the return of those goods!

Comment (mark-prins) [reply client/packages/invoices/src/utils.ts:229]: likely worth adding the check in then - only if manual then allow `Shipped` to be returned. As it is, a customer can return stock that is in transit 🤔

**Review by mark-prins — APPROVED** [review]
had a read through and have run the tests - looks fine to me 🙏 
would be good if someone involved or someone with more rust experience could comment!

File: client/packages/invoices/src/utils.ts (hunk 3 above)
Comment (lache-melvin) [reply client/packages/invoices/src/utils.ts:261]: Curious for more thoughts on this one... this is for processing the return of goods from an outbound shipment - i.e. a supplier saying yes I shipped these goods but now they are being returned. I'd guess that most of the time, inbound returns would be generated from outbound returns (created by the customer), but where you did want to create one manually, would it be worth constraining that you can't do that until the customer marks that shipment as delivered?

**Pull request merged**